apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,378 @@
+	//   - secondary: Ascending order based on the name of the cluster if the label key is absent.
+	// +optional
+	SortingLabelKey *string `json:"sortingLabelKey,omitempty"`
+
+	// The collection of tasks that each stage needs to complete successfully before moving to the next stage.
+	// Each task is executed in parallel and they can not be more than one task of the same type.
+	// +optional
+	AfterStageTasks []AfterStageTask `json:"afterStageTasks,omitempty"`
+}
+
+// AfterStageTask is the collection of post stage tasks that ALL need to be completed before we can move to the next stage.
+type AfterStageTask struct {
+	// The type of the after stage task.
+	// +kubebuilder:validation:Enum=TimedWait;Approval
+	// +required
+	Type AfterStageTaskType `json:"type"`
+
+	// The prefix of the name of the StageApprovalRequest we need to create to get the approval for the next stage.
+	// We create the full name of the approval request by concatenating the prefix to the stagedUpdateRunName and the stageName.
+	// +optional
+	ApprovalRequestPrefix string `json:"approvalTask,omitempty"`
+
+	// The time to wait after all the clusters in the current stage complete the update before we move to the next stage.
+	// +kubebuilder:default:="1h"
+	// +kubebuilder:validation:Pattern="^0|([0-9]+(\\.[0-9]+)?(s|m|h))+$"
+	// +kubebuilder:validation:Type:=string
+	// +optional
+	WaitTime metav1.Duration `json:"waitTime,omitempty"`
+}
+
+// StagedUpdateRunStatus defines the observed state of the StagedUpdateRun.
+type StagedUpdateRunStatus struct {
+	// UpdateRunStrategy is the collection of the stageConfig in the StagedUpdateStrategy that we are going to use for the update.
+	// +required
+	UpdateRunStrategy []StageConfig `json:"updateRunStrategy"`
+
+	// AppliedStrategy is applied strategy of the CRP that we are currently using.
+	// It is the same as the strategy in the CRP when we first start the update.
+	// We will NOT update the strategy in the update even if the appliedStrategy changes in the CRP.
+	// +required
+	AppliedStrategy v1beta1.ApplyStrategy `json:"appliedStrategy"`
+
+	// CurrentStageIndex is the name of the stage that is in the middle of the update.
+	// If we are in the DeletionStage, the name is called "deletionStage".
+	// +required
+	CurrentUpdatingStage string `json:"currentUpdatingStage"`
+
+	// Stages list the current status of each stage.
+	// +required
+	Stages []StageStatus `json:"currentStages"`
+
+	// DeletionStageStatus list the current status of the deletion stage.
+	// +required
+	DeletionStageStatus StageStatus `json:"toBeRevertedClusters"`
+
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	//
+	// Conditions is an array of current observed conditions for StagedUpdateRun.
+	// Known conditions are "RollingOut".
+	// +optional
+	Conditions []metav1.Condition `json:"conditions"`
+}
+
+type StageStatus struct {
```

Comment:
Could you add conditions for the run, each stage, and each cluster? Currently only after stage tasks have conditions.

The condition type could be the same, like pending -> updating (as this is a staged update) -> succeeded/failed.

"CurrentUpdatingStage" is unnecessary if you have conditions per stage.

Reply:
I am not sure if we need conditions inside each stage anymore since the updateStatus  has all the info we need (and  we can add more if needed). Only overall condition missing is in the updateRun itself.